keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,468 @@
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+A developer is adding a new feature of a new write type. Because it requires a consensus write, the request should be rejected if the feature is not enabled or the request can be processed on all cluster members. 
+Today, each change in etcd API, requires development spenning at least 2 minor versions. First a raft proto change needs to be added and wait a release before it can be used. Only a release later it can be actually used to prevent a incompatible proto being used during upgrade/downgrade process. 
+
+With cluster level feature gate, a server would know if a feature is enabled on all members, and decide whether or not to accept the new write request.
+
+#### Story 2
+
+A Kubernetes developer would like use a new great etcd feature, but it is not available on all supported etcd versions. What can they do?
+Historically, Kubernetes avoided using any features that were not available on all supported versions of etcd.
+From K8s 1.31, Kubernetes added very complicated and fragile system: [`FeatureSupportChecker`](https://github.com/kubernetes/kubernetes/blob/db82fd1604ebf327ab74cde0a7158a8d95d46202/staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature/feature_support_checker.go#L42) to detect etcd version, parse it and guess whether this feature based on a set of hardcoded etcd version. It does not really know if the feature is enabled by the feature `--experimental` flag.
+
+With a client-side mechanism to reliably determine if a cluster-wide feature is enabled,  would allow Kubernetes to immediately utilize new features.
+
+#### Story 3
+
+In a HA cluster, users would like to turn off an enabled feature. They need to restart each server node one by one with feature changed from enabled to disabled. Today, after the restarting process begins and before all nodes are restarted, the behavior of the feature is really undefined: it is enabled on some nodes and disabled on the other nodes.
+
+Now with the cluster feature gate, we can deterministically tell if the feature is enabled or disabled for the whole cluster at any moment.
+
+This story applies similar to the upgrade/downgrade scenarios as well.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Could it done on the client side instead of inside etcd? 
+One could argue that it is easy to know if a feature enabled for the whole cluster by querying each member if the feature is enabled on the client side. But there are several caveates with this approach:
+* not all members are available all the time. if some member is not available, it is questionable if the feature can be used, and that could break the HA assumption. 
+* some feature might change how a raft message is sent and applied, and the order of messages relative to the index the feature is enabled is critical. Here the consistent index is more important than the time, and it is hard to get that outside etcd. 
```

Comment:
I think this points is a little unclear. I think it's about fact that order of operations in distributed system can be hard to predict, as they can be arbitrarily re-ordered by leader, which can also be arbitrarily changed. In short, you cannot check if feature is enabled and then assume it will still be enabled when you send a following request.